testmodule/testGasEstimation.nim

```diff
@@ -41,3 +41,37 @@ suite "gas estimation":
     # fails with "Transaction ran out of gas" when gas estimation
     # is not done using the pending block
     await contract.checkTimeEquals(time)
+
+  test "contract gas estimation uses pending block":
+    let latest = CallOverrides(blockTag: some BlockTag.latest)
+    let pending = CallOverrides(blockTag: some BlockTag.pending)
+
+    # retrieve time of pending block
+    let time = await contract.getTime(overrides=pending)
+
+    # ensure that time of latest block and pending block differ
+    check (await contract.getTime(overrides=latest)) != time
+
+    # estimate gas
+    let gas = await contract.estimateGas.checkTimeEquals(time)
+    let overrides = TransactionOverrides(gasLimit: some gas)
+
+    # fails with "Transaction ran out of gas" when gas estimation
+    # was not done using the pending block
+    await contract.checkTimeEquals(time, overrides)
```

Comment:
This is quite confusing to me. The "pending" block is the next block that will be produced and because Hardhat has automine enabled, the next block is produced immediately when a transaction occurs. Therefore, a transaction happens and block should be produced when `checkTimeEquals` is called, so wouldn't that make the `block.timestamp` of the transaction equal to the "pending" block timestamp just before it was called?

Reply:
Updated the comments in 6ef1ce5300b7bac35ccd89a5d1792300f5d6e10e so they are less confusing, I hope.